manager/manager.go

```diff
@@ -0,0 +1,283 @@
+	// unmanaged flytepropeller pods - which is invalid configuration but will be detected later
+
+	// determine missing managed pods
+	podExists := make(map[string]bool)
+	for _, podName := range podNames {
+		podExists[podName] = false
+	}
+
+	podsRunning := 0
+	for _, pod := range pods.Items {
+		podName := pod.ObjectMeta.Name
+		if _, ok := podExists[podName]; ok {
+			podExists[podName] = true
+
+			if pod.Status.Phase == v1.PodRunning {
+				podsRunning++
+			} else if pod.Status.Phase == v1.PodFailed {
+				logger.Warnf(ctx, "flytepropeller pod '%s' in 'failed' state", podName)
+			}
+		} else {
+			logger.Warnf(ctx, "detected unmanaged pod '%s'", podName)
+		}
+	}
+
+	m.metrics.PodsRunning.Set(float64(podsRunning))
+
+	// create non-existent pods
+	for i, podName := range podNames {
+		if exists := podExists[podName]; !exists {
+			pod := m.pod.DeepCopy()
+			pod.ObjectMeta.Name = podName
+
+			err := m.shardStrategy.UpdatePodSpec(&pod.Spec, i)
+			if err != nil {
+				logger.Errorf(ctx, "failed to update pod spec for '%s' [%v]", podName, err)
+				continue
+			}
+
+			_, err = m.kubePodsClient.Create(ctx, pod, metav1.CreateOptions{})
+			if err != nil {
+				logger.Errorf(ctx, "failed to create pod '%s' [%v]", podName, err)
+				continue
+			}
+
+			m.metrics.PodsCreated.Inc()
+			logger.Infof(ctx, "created pod '%s'", podName)
+		}
+	}
+
+	return nil
+}
+
+func (m *Manager) deletePods(ctx context.Context) error {
+	podNames := m.getPodNames()
+	for _, podName := range podNames {
+		err := m.kubePodsClient.Delete(ctx, podName, metav1.DeleteOptions{})
+		if err != nil {
+			if kubeerrors.IsNotFound(err) {
+				logger.Warnf(ctx, "deleting pod '%s' does not exist", podName)
+				continue
+			}
+
+			return err
+		}
+
+		m.metrics.PodsCreated.Inc()
+		logger.Infof(ctx, "deleted pod '%s'", podName)
+	}
+
+	return nil
+}
+
+func (m *Manager) getPodNames() []string {
+	podCount := m.shardStrategy.GetPodCount()
+	var podNames []string
+	for i := 0; i < podCount; i++ {
+		podNames = append(podNames, fmt.Sprintf("%s-%d", m.podApplication, i))
+	}
+
+	return podNames
+}
+
+// Runs either as a leader -if configured- or as a standalone process.
+func (m *Manager) Run(ctx context.Context) error {
+	if m.leaderElector != nil {
+		logger.Infof(ctx, "running with leader election")
+		m.leaderElector.Run(ctx)
+	} else {
+		logger.Infof(ctx, "running without leader election")
+		if err := m.run(ctx); err != nil {
+			return err
+		}
+		m.shutdown()
+	}
+
+	return nil
+}
+
+func (m *Manager) run(ctx context.Context) error {
+	logger.Infof(ctx, "started manager")
+	wait.UntilWithContext(ctx,
+		func(ctx context.Context) {
+			logger.Debugf(ctx, "validating managed pod(s) state")
+			err := m.createPods(ctx)
+			if err != nil {
+				logger.Errorf(ctx, "failed to create pod(s) [%v]", err)
+			}
+		},
+		m.scanInterval,
+	)
+
+	logger.Infof(ctx, "shutting down manager")
+	return nil
+}
+
+func (m *Manager) shutdown() {
```

Comment:
why delete pods on shutdown?

Reply:
also how is the shutdown invoked? We dont want the pods to go down if manager gets a SIGINT

Reply:
if manager 1 starts and uses hash sharding with 4 replicas we have a 32 bit range distributed with 8 for each, the manager fails and flytepropeller-0, flytepropeller-1, 2, and 3 ... are still running. if another managers starts sharding with 8 replicas than it starts flytepropeller-4, 5, 6, 7. however, flytepropeller-3 covers a superset of keys that 4 does similar overlap issues with project and domain sharding. i figured any shutdown on the manager kills all underlying replicas - and that will keep restarts / autoscaling / etc much more clean. maybe something we need to talk about?

Reply:
fixed this per our discussion. we are not storing the pod template resource version and shard strategy hash within each pod, if either changes we delete the pod and restart with the new information.